charms/jupyter-controller/tests/integration/test_charm.py

```diff
@@ -161,3 +166,110 @@ async def test_prometheus_integration(ops_test: OpsTest):
     wait=tenacity.wait_exponential(multiplier=1, min=1, max=10),
     reraise=True,
 )
+
+
+@tenacity.retry(
+    wait=tenacity.wait_exponential(multiplier=2, min=1, max=10),
+    stop=tenacity.stop_after_attempt(30),
+    reraise=True,
+)
+def assert_replicas(client, resource_class, resource_name, namespace):
+    """Test for replicas. Retries multiple times to allow for notebook to be created."""
+
+    dep = client.get(resource_class, resource_name, namespace=namespace)
+    replicas = dep.get("status", {}).get("readyReplicas")
```

Comment:
`readyReplicas==1` is part of what is needed to be ready, but there could be other things we're waiting on too.  The `status.conditions` list has a condition of type `Ready` that has a status field - I think that's the best thing to check.  That is the Notebook Controller's "is everything needed for this notebook ready" report

I copied this status struct from some earlier tests:
```
status:
  conditions:
  - lastProbeTime: "2023-04-04T17:32:38Z"
    lastTransitionTime: "2023-04-04T17:32:38Z"
    status: "True"
    type: Initialized
  - lastProbeTime: "2023-04-04T17:32:38Z"
    lastTransitionTime: "2023-04-04T17:32:32Z"
    message: 'containers with unready status: [mytest istio-proxy]'
    reason: ContainersNotReady
    status: "False"
    type: Ready
  - lastProbeTime: "2023-04-04T17:32:38Z"
    lastTransitionTime: "2023-04-04T17:32:32Z"
    message: 'containers with unready status: [mytest istio-proxy]'
    reason: ContainersNotReady
    status: "False"
    type: ContainersReady
  - lastProbeTime: "2023-04-04T17:32:38Z"
    lastTransitionTime: "2023-04-04T17:32:32Z"
    status: "True"
    type: PodScheduled
  containerState:
    waiting:
      reason: PodInitializing
  readyReplicas: 0
```

It is a little awkward, but you want to do something like find the condition in `status.conditions` that has `type==Ready`, then check whether that condition's `status==True`

Reply:
Will `repliacsReady: 1` indicate that pod is up and running? Looks like type `Ready` cannot be set without non zero replicas.

Reply:
I have to stick with check for replicas, because `conditions` in `status` are not always reliable. They depend on container image.